### PR TITLE
Changing codestyle logic with arrows

### PR DESCRIPTION
**Pull request by akuleshov7**
### What's done:
1) Adding logic of line breaks for lambdas

### CODE REVIEW

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 1]
```diff
@@ -709,28 +709,35 @@ All variants of a (private) val logger should be placed in the beginning of the
...
+            checkSomething()
+         }
+        CLASS -> checkClassElements(node)
+    }
     ```
-  
 **Exception:** *Only* The only exception is ternary operator in Kotlin (it is a single line `if () <> else <>` ) 
 When the entire expression can be 
 
+2) No need to use braces for the body of lambdas and when-conditions (after an arrow)
```
Comment (akuleshov7) [inline info/diktat-kotlin-coding-style-guide-en.md:727]: not supported in diktat - need to create an issue

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 2]
```diff
@@ -833,6 +840,7 @@ Exceptions cases:
 4) Supertype lists: \
    a) 4 spaces if colon before supertype list is on new line \
    b) 4 spaces before each supertype; 8 if colon is on new line
+5) When statements - 8 spaces are used to indent condition branches: (`CONDITION -> foo()`)
```
Comment (akuleshov7) [inline info/diktat-kotlin-coding-style-guide-en.md:843]: not supported in diktat? - need to create an issue

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 1 above)
Comment (petertrr) [reply info/diktat-kotlin-coding-style-guide-en.md:727]: https://github.com/cqfn/diKTat/issues/132

Comment (petertrr) [reply info/diktat-kotlin-coding-style-guide-en.md:727]: Also, I think you should place this requirement after examples of ternary if-else?

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 2 above)
Comment (petertrr) [reply info/diktat-kotlin-coding-style-guide-en.md:843]: Do you mean indentation inside block that follows an arrow, or if single statement starts on a new line? Like
```
condition -> {
        // indent with 8 spaces
        foo()
        bar()
}
```
This case should be covered: both opening braces increase indentation by 4 spaces.

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 3]
```diff
@@ -941,46 +949,57 @@ This international code style prohibits non-latin (non ASCII) symbols in the cod
...
-                .filter(true)
-                .size      
-     ```
-      
-  2. Newline should be placed after assignment operator ('=')
-  3. The name of a function or constructor should not be split by a newline from the opening brace '('. Brace follows immediately after the name without any spaces.   
-  4. Newline should be placed right after the comma (',')
-  5. If lambda expression has one-argument, a newline can appear after an arrow, if an expression has more than one argument before an arrow - breakpoint should appear after the brace.
+1. Compared to Java Kotlin allows not to to put semicolumn (';') after each statement separated by newline.
+ There should be no redundant semicolumn at the end of lines.
```
Comment (petertrr) [inline info/diktat-kotlin-coding-style-guide-en.md:953]:
```suggestion
1. Compared to Java Kotlin allows not to to put semicolon (';') after each statement separated by newline.
 There should be no redundant semicolon at the end of lines.
```

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 4]
```diff
@@ -941,46 +949,57 @@ This international code style prohibits non-latin (non ASCII) symbols in the cod
...
+ 
+ In case when newline is needed to split the line, it should be placed after operators like &&/||/+/e.t.c
+ But newline should be placed before operators like ('.', '?.', '?:', '::', e.t.c), operator !! should not be separated from the value it is checking.
```
Comment (petertrr) [inline info/diktat-kotlin-coding-style-guide-en.md:956]: Also, why isn't `!!` listed with other operators that need newline after them? I don't see any differences in requirements. I guess separating `!!` won't even compile.

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 5]
```diff
@@ -941,46 +949,57 @@ This international code style prohibits non-latin (non ASCII) symbols in the cod
...
+        bar()
+    }
+```
+
+Recommended examples：
+```kotlin
+when (node.elementType) {
+         SOMETHING -> 
+             foo()
+             bar()
```
Comment (petertrr) [inline info/diktat-kotlin-coding-style-guide-en.md:983]: I don't think multiple statements can be not wrapped in braces - at least I couldn't make similar compilable example.

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 6]
```diff
@@ -941,46 +949,57 @@ This international code style prohibits non-latin (non ASCII) symbols in the cod
...
+ In case when newline is needed to split the line, it should be placed after operators like &&/||/+/e.t.c
+ But newline should be placed before operators like ('.', '?.', '?:', '::', e.t.c), operator !! should not be separated from the value it is checking.
+ 
+ Note, that you need to follow functional style:
+ ```kotlin
+    val value = otherValue!!
+            .map(x -> x)
+            .filter(true)
+            .size      
+ ```
```
Comment (petertrr) [inline info/diktat-kotlin-coding-style-guide-en.md:964]: We also wanted to add exception on ternary if-else here, like
```
         if (condition) list.map { foo(it) }.filter { bar(it) } else list.drop(1)
```

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 4 above)
Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:956]:
```
a!! // here you use operator - it is separated from dot, it has absolutely different logic and meaning than `.?` or `.`
    .map{}
```

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 5 above)
Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:983]: yes, true, will fix it now

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 3 above)
Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:953]: fixed

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 2 above)
Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:843]: no, my bad, it is a bug

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 6 above)
Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:964]: added

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 4 above)
Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:955]: `!!` differs completely with unbreakable operators like `?.` as it is not related to `.`

Comment (petertrr) [reply info/diktat-kotlin-coding-style-guide-en.md:955]: Yes, I mean what is negative example in this case? What is 'separated from the value it is checking'? Is it something like
```
a
    !!.map { }
```
?

Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:955]: yes, exactly 
a

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 7]
```diff
@@ -709,28 +709,35 @@ All variants of a (private) val logger should be placed in the beginning of the
...
+2) No need to use braces for the body of lambdas and when-conditions (after an arrow)
+  
 Bad example:
 
 ```kotlin
 val value = if (string.isEmpty())  // WRONG!
                 0
             else
                 1
+
```
Comment (petertrr) [inline info/diktat-kotlin-coding-style-guide-en.md:736]: I still think that mixing examples on lambdas and ternary if-else seems misleading for readers.

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 4 above)
Comment (petertrr) [reply info/diktat-kotlin-coding-style-guide-en.md:955]: As far as I can tell, kotlin doesn't allow to write like this: `!!` can be separated from corresponding object by spaces, not by newlines. This example compiles:
```
a    !!.map { }
```
While your example doesn't.
Or am I missing something here?

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 8]
```diff
@@ -941,46 +948,55 @@ This international code style prohibits non-latin (non ASCII) symbols in the cod
...
+ In case when newline is needed to split the line, it should be placed after operators like &&/||/+/e.t.c
+ But newline should be placed before operators like ('.', '?.', '?:', '::', e.t.c), operator !! should not be separated from the value it is checking.
+ 
+ Note, that you need to follow functional style:
+ ```kotlin
+    val value = otherValue!!
+            .map(x -> x)
+            .filter(true)
+            .size      
+ ```
```
Comment (petertrr) [inline info/diktat-kotlin-coding-style-guide-en.md:963]:
```suggestion
 Note, that you need to follow functional style (i.e. each call with `.` should start at a new line):
 ```kotlin
    val value = otherValue!!
            .map(x -> x)
            .filter {
                true
            }
            .size      
 ```
```
I suggest making this part more verbose and explicitly stating that dot cannot follow closing brace of lambda (for me it was not obvious at first).

File: info/diktat-kotlin-coding-style-guide-en.md (hunk 4 above)
Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:955]: yes, looks that it cannot be compiled, but such code should be fixed to my mind:
```
a!!.map {}.map{}.map{}
```

to this one:
```
a!!
    .map{}
    .map{}
    .map{}
```

Is this the case for you?

Comment (petertrr) [reply info/diktat-kotlin-coding-style-guide-en.md:955]: Yes, sure, it's just that wording was a little misleading at first.

Comment (akuleshov7) [reply info/diktat-kotlin-coding-style-guide-en.md:955]: @petertrr please also be careful that you should not do the following:
```
a!!.map{}
    .map{}
    .map{}
```

**Pull request merged**